mpi-proxy-split/mpi-wrappers/mpi_cart_wrappers.cpp

```diff
@@ -222,6 +161,89 @@ USER_DEFINED_WRAPPER(int, Dims_create, (int)nnodes, (int)ndims, (int *)dims)
   return retval;
 }
 
+#ifdef SINGLE_CART_REORDER
+// This variable holds the cartesian properties and is only used at the time of
+// checkpoint (DMTCP_EVENT_PRECHECKPOINT event in mpi_plugin.cpp).
+CartesianProperties g_cartesian_properties = { .comm_old_size = -1,
+                                               .comm_cart_size = -1,
+                                               .comm_old_rank = -1,
+                                               .comm_cart_rank = -1 };
+
+USER_DEFINED_WRAPPER(int, Cart_create, (MPI_Comm)old_comm, (int)ndims,
+                     (const int *)dims, (const int *)periods, (int)reorder,
+                     (MPI_Comm *)comm_cart)
+{
+  //JASSERT(g_cartesian_properties.comm_old_size == -1)
+  //  .Text("MPI_Cart_create() called more than once. Current implementation "
+  //        "only supports one cartesian communicator.");
+
+  std::function<int()> realBarrierCb = [=]() {
+    int retval;
+    DMTCP_PLUGIN_DISABLE_CKPT();
+    MPI_Comm realComm = VIRTUAL_TO_REAL_COMM(old_comm);
+    JUMP_TO_LOWER_HALF(lh_info.fsaddr);
+    retval = NEXT_FUNC(Cart_create)(realComm, ndims, dims, periods, reorder,
+                                    comm_cart);
+    RETURN_TO_UPPER_HALF();
+    g_cartesian_properties.ndims = ndims;
+    g_cartesian_properties.reorder = reorder;
+    for (int i = 0; i < ndims; i++) {
+      g_cartesian_properties.dimensions[i] = dims[i];
+      g_cartesian_properties.periods[i] = periods[i];
+    }
+    MPI_Comm_size(old_comm, &g_cartesian_properties.comm_old_size);
+    MPI_Comm_size(*comm_cart, &g_cartesian_properties.comm_cart_size);
+    MPI_Comm_rank(old_comm, &g_cartesian_properties.comm_old_rank);
+    MPI_Comm_rank(*comm_cart, &g_cartesian_properties.comm_cart_rank);
+    MPI_Cart_coords(*comm_cart, g_cartesian_properties.comm_cart_rank,
+                    g_cartesian_properties.ndims,
+                    g_cartesian_properties.coordinates);
+
+    if (retval == MPI_SUCCESS && MPI_LOGGING()) {
+      MPI_Comm virtComm = ADD_NEW_COMM(*comm_cart);
+      VirtualGlobalCommId::instance().createGlobalId(virtComm);
+      *comm_cart = virtComm;
+      active_comms.insert(virtComm);
+
+      FncArg ds = CREATE_LOG_BUF(dims, ndims * sizeof(int));
+      FncArg ps = CREATE_LOG_BUF(periods, ndims * sizeof(int));
+      LOG_CALL(restoreCarts, Cart_create, old_comm, ndims, ds, ps, reorder,
+               virtComm);
+    }
+    DMTCP_PLUGIN_ENABLE_CKPT();
+    return retval;
+  };
+  return twoPhaseCommit(old_comm, realBarrierCb);
+}
+#else
+
+USER_DEFINED_WRAPPER(int, Cart_create, (MPI_Comm) old_comm, (int) ndims,
+                     (const int*) dims, (const int*) periods, (int) reorder,
+                     (MPI_Comm *) comm_cart)
+{
+  int retval;
+  reorder = 0;
```

Comment:
Let's add a comment above this explaining why we are enforcing reorder to be false.